### PR TITLE
[RDBMS] Add Test Scenarios

### PR DESCRIPTION
**Pull request by Suna-MS**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [ ] Each command and parameter has a meaningful description.
- [ ] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by tjprescott** [comment]
These have been run live successfully against the staging environment. This PR will be updated with the test recordings once the production environment is live.

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    ScenarioTest)
+from azure.cli.testsdk.preparers import (
+    AbstractPreparer,
+    SingleValueReplacer)
+
+from azure.cli.testsdk.utilities import create_random_name
+
+
+# Constants
+server_name_prefix = 'clitestserver'
```
Comment (troydai) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: All cap for constant: http://legacy.python.org/dev/peps/pep-0008/#constants

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+server_name_max_length = 63
+
+
+class ServerPreparer(AbstractPreparer, SingleValueReplacer):
```
Comment (troydai) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: The name is too generic.

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, engine_type='mysql', engine_parameter_name='engine',
+                 name_prefix=server_name_prefix, parameter_name='server', location='northeurope',
+                 admin_user='cloudsa', admin_password='SecretPassword123',
+                 resource_group_parameter_name='resource_group', skip_delete=False):
```
Comment (troydai) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: If deleting the resource group will delete the server, make skip_delete to be `True` actually saves time during live test.

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+# Constants
+server_name_prefix = 'clitestserver'
+server_name_max_length = 63
+
+
+class ServerPreparer(AbstractPreparer, SingleValueReplacer):
+    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, engine_type='mysql', engine_parameter_name='engine',
```
Comment (troydai) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: `engine` is a generic name, it is recommended to be a more specific name. for example, `database_engine`

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+                              JMESPathCheck('type(@)', 'array')]).get_output_in_json()
+
+        # test download file
+        file_name = result[0]['name']
+        self.cmd('{} server-logs download -g {} -s {} --name {}'
+                 .format(engine, rg, server, file_name),
+                 checks=[NoneCheck()])
+
+        import os
+        os.remove(file_name)
```
Comment (troydai) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: Use `addCleanup` instead to make sure the file is deleted even if the assertion fails. https://docs.python.org/2/library/unittest.html#unittest.TestCase.addCleanup

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+from azure.cli.testsdk.utilities import create_random_name
+
+
+# Constants
+server_name_prefix = 'clitestserver'
+server_name_max_length = 63
+
+
+class ServerPreparer(AbstractPreparer, SingleValueReplacer):
+    # pylint: disable=too-many-arguments
```
Comment (troydai) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: I just disable the `too-many-arguments` rule project-wise. So you don't need this anymore.

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py (hunk 1 above)
Comment (Suna-MS) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: Will resolve in the commit

**Comment by troydai** [comment]
Taking a second look.

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+            execute('az {} server delete -g {} -n {} --yes'.format(self.engine_type, group, name))
+
+    def _get_resource_group(self, **kwargs):
+        try:
+            return kwargs.get(self.resource_group_parameter_name)
+        except KeyError:
+            template = 'To create a rdbms server, resource group is required. Please add ' \
+                       'decorator @{} in front of this server preparer.'
+            raise CliTestError(template.format(ResourceGroupPreparer.__name__,
+                                               self.resource_group_parameter_name))
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: I may have missed something obvious but `template` has one parameter but 2 params are passed to `template.format(...)`.

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+                 checks=[
+                     JMESPathCheck('name', servers[0]),
+                     JMESPathCheck('resourceGroup', rg),
+                     JMESPathCheck('sslEnforcement', 'Enabled'),
+                     JMESPathCheck('tags.key', '3'),
+                     JMESPathCheck('sku.tier', edition),
+                     JMESPathCheck('administratorLogin', admin_login)])
+
+        # test restore to a new server, make sure wait at least 5 min after server created.
+        from time import sleep
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: For clarity, this import should just be at the top of the file.

File: src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py (hunk 7 above)
Comment (Suna-MS) [inline src/command_modules/azure-cli-rdbms/tests/test_rdbms_commands.py]: Confirmed that dict.get will not throw KeyError exception so removed the statement.

**Pull request merged**